File: 18_enums/enums.rb
# ENUM 1: Integer enum using constants
class Status
  PENDING  = 0
  APPROVED = 1
  REJECTED = 2
end

# ENUM 2: String enum (API friendly)
class PaymentMethod
  CREDIT_CARD = "credit_card"
  UPI         = "upi"
  NET_BANKING = "net_banking"
end

# ENUM 3: Hash-based enum (frozen for safety)
ROLE = {
  admin: "ADMIN",
  user: "USER",
  guest: "GUEST"
}.freeze

# ENUM validation
def valid_status?(status)
  [Status::PENDING, Status::APPROVED, Status::REJECTED].include?(status)
end

# -------- Usage (single flow) --------
status = Status::APPROVED
payment = PaymentMethod::UPI
role = ROLE[:admin]

puts "Status: #{status}"
puts "Valid Status? #{valid_status?(status)}"
puts "Payment Method: #{payment}"
puts "Role: #{role}"

File: 18_enums/enums.go
package main

import "fmt"

/*
	ENUM 1: Integer enum using iota
*/
type Status int

const (
	Pending Status = iota
	Approved
	Rejected
)

/*
	ENUM 2: Enum with custom integer values
*/
type Role int

const (
	Admin Role = 1
	User  Role = 2
	Guest Role = 3
)

/*
	ENUM 3: Enum with String() method
*/
type OrderStatus int

const (
	Created OrderStatus = iota
	Shipped
	Delivered
	Cancelled
)

func (o OrderStatus) String() string {
	return [...]string{
		"CREATED",
		"SHIPPED",
		"DELIVERED",
		"CANCELLED",
	}[o]
}

/*
	ENUM 4: String-based enum (Best for APIs)
*/
type PaymentMethod string

const (
	CreditCard PaymentMethod = "CREDIT_CARD"
	UPI        PaymentMethod = "UPI"
	NetBanking PaymentMethod = "NET_BANKING"
)

/*
	ENUM Validation
*/
func IsValidStatus(s Status) bool {
	switch s {
	case Pending, Approved, Rejected:
		return true
	default:
		return false
	}
}

func main() {

	// Integer enum
	var s Status = Approved
	fmt.Println("Status:", s)
	fmt.Println("Valid Status?", IsValidStatus(s))

	// Custom value enum
	var r Role = Admin
	fmt.Println("Role:", r)

	// Enum with String method
	var o OrderStatus = Shipped
	fmt.Println("Order Status:", o)

	// String enum
	var p PaymentMethod = UPI
	fmt.Println("Payment Method:", p)
}
